File: rxutil/src/main/java/com/inqbarna/rxutil/paging/RxPagingAdapterDelegate.java
package com.inqbarna.rxutil.paging;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;

import com.inqbarna.common.paging.PaginatedAdapterDelegate;
import com.inqbarna.common.paging.PaginatedList;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func0;
import rx.functions.Func3;
import rx.observables.AsyncOnSubscribe;

/**
 * @author David García <david.garcia@inqbarna.com>
 * @version 1.0 16/9/16
 */

public class RxPagingAdapterDelegate<T> extends PaginatedAdapterDelegate<T> {

    private RxPaginatedList.Callbacks mCallbacks = new RxPaginatedList.Callbacks() {
        @Override
        public void onError(Throwable throwable) {
            mErrorCallback.onError(throwable);
        }

        @Override
        public void onItemsAdded(int startPos, int size) {
            onFinishAddItems(startPos, size);
        }
    };

    private ErrorCallback mErrorCallback;
    private Subscription mActiveSubscription;

    public RxPagingAdapterDelegate(RecyclerView.Adapter adapter, @NonNull ErrorCallback errorCallback, @Nullable ProgressHintListener loadingListener) {
        super(adapter, loadingListener);
        mErrorCallback = errorCallback;
    }

    @Override
    public void setItemsInternal(PaginatedList<T> items, boolean endLoad) {
        setDataFactory(asPageFactory(items), items.size(), items.size(), endLoad);
    }

    private PageFactory<T> asPageFactory(PaginatedList<T> items) {
        final List<T> allItems = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            allItems.add(items.get(i));
        }

        return new PageFactory<T>() {
            @Override
            public Observable<? extends T> nextPageObservable(int start, int size) {
                List<T> subList = allItems.subList(start, start + Math.min(size, allItems.size()));
                return Observable.from(subList);
            }
        };
    }

    public void setDataFactory(PageFactory<T> factory, int displayPageSize, int requestPageSize) {
        setDataFactory(factory, displayPageSize, requestPageSize, false);
    }

    private void setDataFactory(PageFactory<T> factory, int displayPageSize, int requestPageSize, boolean endLoad) {
        setDataStream(createStreamObservable(factory, displayPageSize, requestPageSize), endLoad);
    }

    private Observable<? extends List<? extends T>> createStreamObservable(PageFactory<T> factory, int displayPageSize, int requestPageSize) {
        return createStream(factory, displayPageSize, requestPageSize);
    }

    private Observable<List<T>> createStream(final PageFactory<T> factory, final int displayPageSize, final int requestPageSize) {
        return Observable.create(
                AsyncOnSubscribe.createStateful(
                        new Func0<RequestState<T>>() {
                            @Override
                            public RequestState<T> call() {
                                return new RequestState<>(requestPageSize, factory);
                            }
                        },
                        new Func3<RequestState<T>, Long, Observer<Observable<? extends T>>, RequestState<T>>() {
                            @Override
                            public RequestState<T> call(RequestState<T> state, Long aLong, Observer<Observable<? extends T>> observableObserver) {
                                Throwable error = state.getError();
                                if (null != error) {
                                    observableObserver.onError(error);
                                } else if (state.getCompleted()) {
                                    observableObserver.onCompleted();
                                } else {
                                    Observable<? extends T> ob = state.nextObservable();
                                    observableObserver.onNext(ob);
                                }
                                return state;
                            }
                        }
                )
        ).buffer(displayPageSize).observeOn(AndroidSchedulers.mainThread(), 1);
    }

    private void setDataStream(Observable<? extends List<? extends T>> stream, boolean endLoad) {
        if (null != mActiveSubscription) {
            mActiveSubscription.unsubscribe();
        }
        final PaginatedList<T> items = RxPaginatedList.create(stream, mCallbacks);
        mActiveSubscription = ((RxPaginatedList<T>)items);
        beginProgress();
        super.setItemsInternal(items, endLoad);
    }

    @Override
    public void addNextPage(Collection<? extends T> pageItems, boolean lastPage) {
        throw new UnsupportedOperationException("This delegate does not support this operation");
    }

    public void setDataFactory(PageFactory<T> factory, int pageSize) {
        setDataFactory(factory, pageSize, pageSize);
    }
}

File: rxutil-rxjava2/src/main/java/com/inqbarna/rxutil/paging/RxPagingAdapterDelegate.kt
package com.inqbarna.rxutil.paging

import android.support.v7.widget.RecyclerView
import com.inqbarna.common.paging.PaginatedAdapterDelegate
import com.inqbarna.common.paging.PaginatedList
import io.reactivex.Emitter
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiConsumer
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import org.reactivestreams.Publisher
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

/**
 * @author David García <david.garcia@inqbarna.com>
 * @version 1.0 16/9/16
 */

class RxPagingAdapterDelegate<T>(
        adapter: RecyclerView.Adapter<*>,
        private val mRxPagingCallback: RxPagingCallback,
        paginateConfig: RxPagingConfig,
        loadingListener: PaginatedAdapterDelegate.ProgressHintListener?
) : PaginatedAdapterDelegate<T>(adapter, loadingListener, paginateConfig) {

    private val callbacks = object : RxPaginatedList.Callbacks {
        override fun onError(throwable: Throwable) {
            mRxPagingCallback.onError(throwable)
            endProgress()
        }

        override fun onCompleted() {
            mRxPagingCallback.onCompleted()
            endProgress()
        }

        override fun onItemsAdded(startPos: Int, size: Int) {
            val pgConfig = getPaginateConfig() as RxPagingConfig
            if (pgConfig.notifyAsInsertions) {
                onFinishAddItems(startPos, size)
            } else {
                endProgress()
            }
        }
    }
    private var activeDisposable: Disposable? = null

    override fun setItemsInternal(items: PaginatedList<T>, endLoad: Boolean) {
        setDataFactory(asPageFactory(items), items.size(), items.size(), endLoad)
    }

    private fun asPageFactory(items: PaginatedList<T>): PageFactory<T> {
        val allItems = ArrayList<T>()
        for (i in 0 until items.size()) {
            allItems.add(items.get(i))
        }

        return object : PageFactory<T> {
            override val initialData: List<T> get() = allItems
            override fun nextPageObservable(startOffset: Int, pageSize: Int): Publisher<out T> = Flowable.empty()
        }
    }

    fun setDataFactory(factory: PageFactory<T>, displayPageSize: Int, requestPageSize: Int) {
        setDataFactory(factory, displayPageSize, requestPageSize, factory.initialData.isNotEmpty())
    }

    private fun setDataFactory(factory: PageFactory<T>, displayPageSize: Int, requestPageSize: Int, endLoad: Boolean) {
        setDataStream(createStreamObservable(factory, displayPageSize, requestPageSize), endLoad)
    }

    private fun createStreamObservable(factory: PageFactory<T>, displayPageSize: Int, requestPageSize: Int): Flowable<out List<T>> {
        return createStream(factory, displayPageSize, requestPageSize)
    }

    private fun createStream(factory: PageFactory<T>, displayPageSize: Int, requestPageSize: Int): Flowable<List<T>> {
        return Flowable.generate(
                Callable<RequestState<T>> { RequestState(requestPageSize, factory, paginateConfig as RxPagingConfig) },
                BiConsumer<RequestState<T>, Emitter<T>> { state, emitter ->
                    var nextItem: T?

                    acquireNext@ while (true) {
                        nextItem = state.blockingNext()

                        // Important, we read state after call to blockingNext, as it does change state there!
                        val loadState = state.state

                        when (loadState) {
                            is Error -> {
                                if (loadState.hasRecovery) {
                                    loadState.recovery?.setCallbacks(
                                            object : RetryCallbacks {
                                                override fun onRetryRequested() {
                                                    postAction { onRecoveryInProgress(true) }
                                                }

                                                override fun onRetryAborted() {
                                                    postAction { onRecoveryInProgress(false) }
                                                }
                                            }
                                    )
                                    postAction { disableProgressAwaitForRecovery() }
                                } else {
                                    emitter.onError(loadState.error)
                                    return@BiConsumer
                                }
                            }
                            Complete -> {
                                emitter.onComplete()
                                return@BiConsumer
                            }
                            else -> break@acquireNext
                        }
                    }

                    if (null == nextItem) {
                        throw IllegalStateException("We shouldn't reach this point if value is null")
                    }

                    emitter.onNext(nextItem)
                },
                Consumer { s -> s.dispose() }
        ).subscribeOn(Schedulers.from(PAGE_PREFETCH)).buffer(displayPageSize).observeOn(AndroidSchedulers.mainThread(), false, 1)
    }

    private fun postAction(action: () -> Unit) {
        AndroidSchedulers.mainThread().scheduleDirect(action)
    }

    private fun setDataStream(stream: Flowable<out List<T>>, endLoad: Boolean) {
        activeDisposable?.dispose()
        val items = RxPaginatedList.create(stream, callbacks, paginateConfig as RxPagingConfig)
        activeDisposable = items as RxPaginatedList<T>
        beginProgress()
        super.setItemsInternal(items, endLoad)
    }

    override fun addNextPage(pageItems: Collection<T>, lastPage: Boolean) {
        throw UnsupportedOperationException("This delegate does not support this operation")
    }

    fun setDataFactory(factory: PageFactory<T>, pageSize: Int) {
        setDataFactory(factory, pageSize, pageSize)
    }

    companion object {
        private val counter: AtomicInteger = AtomicInteger(0)
        private val PAGE_PREFETCH: ExecutorService = Executors.newCachedThreadPool { Thread(it, nextName()) }
        private fun nextName(): String {
            return "PagePrefetcher-${counter.getAndIncrement()}"
        }
    }

}
